### PR TITLE
Feature/invitation manager

### LINKED ISSUE #467: Invitation manager
## Invitation email

- [ ] Update the URL on the call to action behind the `JOIN NOW` button: `https://labelflow.ai/[workspace-slug]/accept-invite?invitationToken=yy`
  - [ ] invitation token is a uuid

## Invitation manager

- [ ] Create following pages
  - [ ] Accept/Decline membership page.
  - [ ] This invitation is not valid
  - [ ] This invitation was already accepted
- [ ] Implement invitation manager state diagram [here in figma](https://www.figma.com/file/56hJc9jl9oedCWIfxj7TyU/State-Diagram?node-id=0%3A1)

## State diagram

image

## Mockup
[Accept/Decline invitation](https://www.figma.com/file/4J9xyfF5v4B98MuKVO0Y8V/LabelFlow-mockups?node-id=3317%3A18876)
[Revoked invitation](https://www.figma.com/file/4J9xyfF5v4B98MuKVO0Y8V/LabelFlow-mockups?node-id=3317%3A18877)
[Already accepted](https://www.figma.com/file/4J9xyfF5v4B98MuKVO0Y8V/LabelFlow-mockups?node-id=3318%3A19010)

  
Blocked by #463
Blocked by #461

### PR DESCRIPTION
**Pull request by RaynalHugo**
# Feature

## Work performed

- Added needed backend resolvers to retrieve the `status` and `currentUserCanAcceptInvitation` of a membership.
- Used `status` in the list of members of a workspace rather than a frontend computed value.
- Refactored the invitation process to send an email with a link containing the id of the membership rather than an invitation token.
- Exposed publicly memberships when queried by id to make it possible to retrieve their status.
- Implemented the `/[workspaceSlug]/accept-invite` page

TODO:
- Add `notImplemented` for the resolvers in the SW resolvers.
- Edit email message to remove "will expire in 24h"?
- Handle case where no query param is passed?
- rename by "pending" or "sent" in memberships lists of workspace

## Results

- Accept or decline dialog
image

- Accepted
image

- Declined
image

- Already accepted
image

- Already declined
image

- Already member of the workspace
image

- You need to sign in 
Note that the modal is opened by default when you reach this page without being logged in. This dialog is what is displayed if the user closes the log-in modal. 
image

- Membership doesn't exist
image

## Problems encountered
- The issue was a bit fuzzy but we clarified it together.

## Caveats
- Memberships are now public when queried by id. See discussion: https://labelflow.slack.com/archives/C022PCC0H0W/p1635946220013100

## Resolved issues
closes #467

### CODE REVIEW

**Comment by nicOwlas** [comment]
Looks good! Remarks:
- Already declined. Add after the sentence: "If it wasn't by you, contact your Workspace Administrator."
- Revoked: is there a way to revoke someone?

**Comment by RaynalHugo** [comment]
> * Already declined. Add after the sentence: "If it wasn't by you, contact your Workspace Administrator."

Sure!

> * Revoked: is there a way to revoke someone?

On the membership table, you can delete a membership. If you do this while it wasn't yet accepted, you would end up in this state.
It is also possible to reach this state if you mess with the URL and change the membership id the query parameters.

File: typescript/web/src/pages/[workspaceSlug]/members/index.tsx [hunk 1]
```diff
@@ -104,9 +104,9 @@ const WorkspaceMembersPage = () => {
             }}
             inviteMember={async (where) => {
               const {
-                data: { inviteMember: invitationStatus },
+                data: { inviteMember: InvitationResult },
```
Comment (crubier) [inline typescript/web/src/pages/[workspaceSlug]/members/index.tsx:107]:
```suggestion
                data: { inviteMember: invitationResult },
```

**Review by crubier — APPROVED** [review]

**Review by crubier — APPROVED** [review]

File: cypress/integration/invitation-manager.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+      cy.task("inviteUser").then(({ inviteUrl }: any) => {
+        cy.task("performLogin", {
+          name: "UserNotAlreadyAMember",
+          email: "test+1@labelflow.ai",
+        }).then((token) => {
+          cy.setCookie("next-auth.session-token", token as string);
+          cy.visit(inviteUrl);
+        });
+      });
+
```
[the remark this answers is not in the document]
Comment (RaynalHugo) [reply cypress/integration/invitation-manager.ts:20]: Done in f3c35990bd9ed8d9a8514ea6aae2bacb0cb8bb4a

**Pull request merged**